File: app/src/main/java/com/alvinhkh/buseta/mtr/ui/MtrLineStatusFragment.java
package com.alvinhkh.buseta.mtr.ui;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.alvinhkh.buseta.R;
import com.alvinhkh.buseta.datagovhk.model.MtrLineName;
import com.alvinhkh.buseta.mtr.MtrService;
import com.alvinhkh.buseta.mtr.model.MtrLineStatus;
import com.alvinhkh.buseta.mtr.model.MtrLineStatusRes;
import com.alvinhkh.buseta.ui.ArrayListRecyclerViewAdapter.Item;
import com.google.gson.Gson;

import java.io.IOException;
import java.io.InputStream;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;


public class MtrLineStatusFragment extends Fragment
        implements SwipeRefreshLayout.OnRefreshListener {

    private final MtrService mtrService = MtrService.Companion.getTnews().create(MtrService.class);

    private final CompositeDisposable disposables = new CompositeDisposable();

    private SwipeRefreshLayout swipeRefreshLayout;

    private MtrLineStatusAdapter adapter;

    private FloatingActionButton fab;

    private RecyclerView recyclerView;

    private View emptyView;

    private Map<String, MtrLineName> codeMap = new HashMap<>();

    public MtrLineStatusFragment() {
    }

    /**
     * Returns a new instance of this fragment for the given section
     * number.
     */
    public static MtrLineStatusFragment newInstance() {
        return new MtrLineStatusFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        setHasOptionsMenu(true);
        View rootView = inflater.inflate(R.layout.fragment_list, container, false);
        recyclerView = rootView.findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setHasFixedSize(true);
        adapter = new MtrLineStatusAdapter(recyclerView);
        recyclerView.setAdapter(adapter);
        emptyView = rootView.findViewById(R.id.empty_view);
        TextView emptyText = rootView.findViewById(R.id.empty_text);
        emptyText.setText(R.string.message_no_data);
        if (adapter.getDataItemCount() > 0) {
            recyclerView.setVisibility(View.VISIBLE);
            emptyView.setVisibility(View.GONE);
        } else {
            recyclerView.setVisibility(View.INVISIBLE);
            emptyView.setVisibility(View.VISIBLE);
        }
        swipeRefreshLayout = rootView.findViewById(R.id.swipe_refresh_layout);
        swipeRefreshLayout.setOnRefreshListener(this);

        fab = rootView.findViewById(R.id.fab);
        return rootView;
    }

    @Override
    public void onResume() {
        super.onResume();
        if (fab != null) {
            fab.hide();
        }
        if (swipeRefreshLayout != null) {
            swipeRefreshLayout.setEnabled(false);
            swipeRefreshLayout.setRefreshing(false);
        }
        if (getActivity() != null) {
            FloatingActionButton fab = getActivity().findViewById(R.id.fab);
            if (fab != null) {
                fab.hide();
            }
        }
        onRefresh();
    }

    @Override
    public void onRefresh() {
        if (swipeRefreshLayout != null) {
            swipeRefreshLayout.setRefreshing(true);
        }
        if (adapter != null) {
            adapter.clear();
        }
        codeMap.clear();
        if (getActivity() != null && getActivity().getAssets() != null) try {
            InputStream inputStream = getActivity().getAssets().open("mtr_lines_name.json");
            int size = inputStream.available();
            byte[] buffer = new byte[size];
            inputStream.read(buffer);
            inputStream.close();
            String json = new String(buffer, "UTF-8");
            Gson gson = new Gson();
            MtrLineName[] mtrLineNameArray = gson.fromJson(json, MtrLineName[].class);
            List<MtrLineName> mtrLineNames = Arrays.asList(mtrLineNameArray);
            for (MtrLineName mtrLineName : mtrLineNames) {
                codeMap.put(mtrLineName.getLineCode(), mtrLineName);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        disposables.add(mtrService.lineStatus()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(lineStatusObserver()));
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        switch (id) {
            case R.id.action_refresh:
                onRefresh();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onDestroyView() {
        disposables.clear();
        super.onDestroyView();
    }

    DisposableObserver<MtrLineStatusRes> lineStatusObserver() {
        return new DisposableObserver<MtrLineStatusRes>() {
            @Override
            public void onNext(MtrLineStatusRes res) {
                if (res == null) return;
                if (swipeRefreshLayout != null) {
                    swipeRefreshLayout.setRefreshing(true);
                }
                if (adapter != null) {
                    if (res.getLines() != null) {
                        for (MtrLineStatus status : res.getLines()) {
                            status.setLineName(status.getLineCode());
                            if (!codeMap.containsKey(status.getLineCode())) continue;
                            MtrLineName mtrLineName = codeMap.get(status.getLineCode());
                            status.setLineName(mtrLineName.getNameTc());
                            status.setLineColour(mtrLineName.getColour());
                            adapter.add(new Item(Item.TYPE_DATA, status));
                        }
                    }
                }
            }

            @Override
            public void onError(Throwable e) {
                Timber.d(e);
                if (swipeRefreshLayout != null) {
                    swipeRefreshLayout.setRefreshing(false);
                }
            }

            @Override
            public void onComplete() {
                if (swipeRefreshLayout != null) {
                    swipeRefreshLayout.setRefreshing(false);
                }
                if (adapter != null) {
                    adapter.setLoaded();
                    if (recyclerView != null) {
                        if (adapter.getDataItemCount() > 0) {
                            recyclerView.setVisibility(View.VISIBLE);
                        } else {
                            recyclerView.setVisibility(View.INVISIBLE);
                        }
                    }
                    if (emptyView != null) {
                        if (adapter.getDataItemCount() > 0) {
                            emptyView.setVisibility(View.GONE);
                        } else {
                            emptyView.setVisibility(View.VISIBLE);
                        }
                    }
                }
            }
        };
    }

}

File: app/src/main/java/com/alvinhkh/buseta/mtr/ui/MtrLineStatusFragment.kt
package com.alvinhkh.buseta.mtr.ui

import android.content.Intent
import androidx.lifecycle.ViewModelProvider
import android.os.Bundle
import android.view.*
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.snackbar.Snackbar
import androidx.fragment.app.Fragment
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.alvinhkh.buseta.R
import com.alvinhkh.buseta.ui.image.ImageActivity
import com.alvinhkh.buseta.utils.ConnectivityUtil


class MtrLineStatusFragment: Fragment(), SwipeRefreshLayout.OnRefreshListener {

    private lateinit var viewAdapter: MtrLineStatusViewAdapter
    private lateinit var swipeRefreshLayout: SwipeRefreshLayout
    private lateinit var recyclerView: RecyclerView
    private lateinit var emptyView: View
    private lateinit var snackbar: Snackbar
    private var fab: FloatingActionButton? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.fragment_list, container, false)
        setHasOptionsMenu(true)
        fab = activity?.findViewById(R.id.fab)
        snackbar = Snackbar.make(rootView?.findViewById(R.id.coordinator_layout)?:rootView, R.string.message_no_internet_connection, Snackbar.LENGTH_INDEFINITE)
        swipeRefreshLayout = rootView.findViewById(R.id.swipe_refresh_layout)
        swipeRefreshLayout.setOnRefreshListener(this)
        emptyView = rootView.findViewById(R.id.empty_view)
        emptyView.visibility = View.VISIBLE
        val emptyTextView = rootView.findViewById<TextView>(R.id.empty_text)
        emptyTextView.text = getString(R.string.updating)
        recyclerView = rootView.findViewById(R.id.recycler_view)
        viewAdapter = MtrLineStatusViewAdapter()
        swipeRefreshLayout.isRefreshing = true
        recyclerView.run {
            setHasFixedSize(true)
            layoutManager = LinearLayoutManager(context)
            adapter = viewAdapter
        }
        val lineStatusViewModel = ViewModelProvider(this).get(MtrLineStatusViewModel::class.java)
        lineStatusViewModel.getAsLiveData().observe(viewLifecycleOwner, { list ->
            if (ConnectivityUtil.isConnected(context)) {
                snackbar.dismiss()
            } else {
                emptyTextView.text = getString(R.string.message_no_data)
                snackbar.show()
            }
            if (!swipeRefreshLayout.isRefreshing) {
                swipeRefreshLayout.isRefreshing = true
            }
            viewAdapter.replace(list?: listOf())
            emptyView.visibility = if (viewAdapter.itemCount > 0) View.GONE else View.VISIBLE
            if (swipeRefreshLayout.isRefreshing) {
                swipeRefreshLayout.isRefreshing = false
            }
        })
        val latestAlertViewModel = ViewModelProvider(this).get(MtrLatestAlertViewModel::class.java)
        latestAlertViewModel.getAsLiveData().observe(viewLifecycleOwner, { list ->
            viewAdapter.alert(list?: listOf())
        })
        onRefresh()
        return rootView
    }

    override fun onResume() {
        super.onResume()
        fab?.hide()
        swipeRefreshLayout.isEnabled = false
        swipeRefreshLayout.isRefreshing = false
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.menu_route, menu)
        menu.findItem(R.id.action_search)?.isVisible = false
        menu.findItem(R.id.action_search_open)?.isVisible = false
        menu.findItem(R.id.action_show_map)?.isVisible = true
        menu.findItem(R.id.action_notice)?.isVisible = false
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_show_map -> {
                val intent = Intent(context, ImageActivity::class.java)
                intent.putExtra(ImageActivity.IMAGE_TITLE, "港鐵車務狀況 (運輸署)")
                intent.putExtra(ImageActivity.IMAGE_URL, "http://210.3.170.180/mtr_status/MTR.jpg")
                intent.putExtra(ImageActivity.COLOUR, ContextCompat.getColor(requireContext(), R.color.black))
                startActivity(intent)
                return true
            }
            R.id.action_refresh -> onRefresh()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onRefresh() {
        if (ConnectivityUtil.isConnected(context)) {
            snackbar.dismiss()
        } else {
            snackbar.show()
        }
//        WorkManager.getInstance().enqueue(
//                OneTimeWorkRequest.Builder(MtrTrainSpecialNewsWorker::class.java)
//                        .addTag("MtrTrainSpecialNews").build())
//        WorkManager.getInstance().enqueue(OneTimeWorkRequest.Builder(MtrLineWorker::class.java).addTag("RouteList").build())
    }
}
